[tests][mobile] Enable diagnostic_tracing for mobile runtime tests

mdh1418 opened the pull request:
Before merging this, dotnet/diagnostics needs to be updated to accept a string parameter to the DiagnosticsClient constructor and allow for TCPIP connection. After which, `MicrosoftDiagnosticsNETCoreClientVersion` can be bumped. The tests work locally because a modified manual build of dotnet/diagnostics was used.

This PR looks to stand-up diagnostics tests for mobile, specifically targeting Android. Diagnostics runtime components have been made compatible for mobile platforms, and here, they will be tested and integrated.

The following are changes made:
Includes `diagnostics_tracing` runtime components in `BuildAndroidApp`
Sets the `DiagnosticsPorts` to default to `127.0.0.1:9000,nosuspend,listen`
Fix IpcUtils `GetStandardTransport` to set up TCP/IP connection on Android, connecting to default DiagnosticsPorts
Fixes #52763
Disables `JitOptimizationSensitive` on Android for eventpipe tests

----
With a goal of emulating the diagnostics tracing tests on mono desktop, the following tests now pass locally.

## eventlistener
eventlistener

## eventcounter
eventcounter
pollingcounter
regression-46938
regression-25709
incrementingpollingcounter
runtimecounters
incrementingeventcounter
gh53564

## eventactivityidcontrol
eventactivityidcontrol

----
Of the eventpipe tests, 5 of them will be enabled in a separate PR, and the rest require a significant modification to allow Android to run multiple apps at a time, specifically to allow for RunSubprocess in IpcUtils to work.

name_config_with_pid, reverse, reverseouter, pauseonstart, diagnosticport, processenvironment

fanyang-mono commented:
With Steve's latest change to reduce the amount of runtime tests running per PR, I believe that all runtime tests running on mobile targets have been moved to rolling build. You would need to modify the runtime.yml and/or runtime-stage.yml files to validate your change here.

Commit:
[mobile] Add diagnostics tracing to Android iOS mobile runtime components

Commit:
[tests] Reenable eventactivityidcontrol test on Android

Reenables eventactivityidcontrol now that diagnostics_tracing runtime component is included

Commit:
[tests] Enable non-eventpipe Android tracing tests to emulate desktop mono

mdh1418 commented:
/azp run runtime

lateralusX commented on src/tests/tracing/eventpipe/common/IpcUtils.cs:
```diff
@@ -446,6 +446,12 @@ public static Stream GetStandardTransport(int processId)
+            else if (OperatingSystem.IsAndroid())
+            {
+                Int32 port = 9000;
+                TcpClient client = new TcpClient("127.0.0.1", port);
```
Android emulator doesn't use 127.0.0.1 as its IP (unless you have setup port forwarding using adb). Is that the case? If not, then emulator uses 10.0.2.2 as its loopback interface IP address.

mdh1418 replied:
I has tried using 10.0.2.2 here, but `System.Net.Internals.SocketExceptionFactory+ExtendedSocketException (111): Connection refused 10.0.2.2:9000` is hit. I was guessing that this needs to be the same value as `DiagnosticsPort` set in `build.proj`, and my initial attempts to use `10.0.2.2` didn't work there either.
Should `processinfo.cs` fail here when using `GetStandardTransport`?

lateralusX replied:
We run the client directly on the device for this scenario, so it will be good.

lateralusX reviewed (commented):
Once this builds and pass, could we also validate that the tests are actually executed on CI as expected?

steveisok commented:
/azp run runtime-manual

steveisok commented on src/tests/tracing/eventpipe/common/IpcUtils.cs:
```diff
@@ -446,6 +446,12 @@ public static Stream GetStandardTransport(int processId)
             }
+            else if (OperatingSystem.IsAndroid())
+            {
+                Int32 port = 9000;
```
@akoeplinger double checking with you.  I *think* we're good.

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 commented:
Can I get another look over this PR @lateralusX, all the tests in the updated PR description are passing on CI.

lateralusX commented on src/tests/tracing/eventcounter/eventcounter.csproj:
```diff
@@ -4,7 +4,7 @@
     <GCStressIncompatible>true</GCStressIncompatible>
     <!-- This test is timing sensitive and JIT timing affects the results of the test -->
-    <JitOptimizationSensitive>true</JitOptimizationSensitive>
+    <JitOptimizationSensitive Condition="'$(TargetOS)' != 'Android'">true</JitOptimizationSensitive>
```
Is this just to get the test running on the Android CoreCLR runtime test configuration? Believe this should be set when we run the runtime tests used the same configuration as we do on OSX, it should set the COM_TieredCompilation=0 as part of running no_tiered_compilation, that in turn should be part of the innerloop configuration, https://github.com/dotnet/runtime/blob/4c662b16e3e6fe688f9bbff2a9423fdee799ecfe/eng/pipelines/runtime.yml#L1062. Similar to what we did in this PR, https://github.com/dotnet/runtime/pull/58106

Having that said it might be equal important to be less dependent on all these configs and since this is not an issue on MonoVM (not running tiered JIT), we could just make sure we run the tests independently.

mdh1418 replied:
Good catch. Now that I'm inspecting this, I do believe I had been running the test on Android CoreCLR runtime test configuration. I thought it had been necessary to exclude `JitOptimizationSensitive` to run the test, but I guess its a CoreCLR specific property.

I thought I was running on mono runtime with 
`<reporoot>/build.sh -s mono+libs -os Android -arch x64 -c Debug`
`<reporoot>/src/tests/build.sh /p:LibrariesConfiguration=Debug -excludemonofailures -rebuild mono -os Android debug -test:tracing/eventcounter/eventcounter.csproj`
`<reporoot>/dotnet.sh artifacts/tests/coreclr/Android.x64.Debug/Tests/Core_Root/xunit.console.dll artifacts/tests/coreclr/Android.x64.Debug/tracing/eventcounter/tracing.eventcounter.XUnitWrapper.dll`
But this is hitting `SKIPPING EXECUTION BECAUSE COMPlus_TieredCompilation has not been disabled and this test is marked JitOptimizationSensitive
Test Harness Exitcode is : 0` so it must still be CoreCLR?

I did see `RuntimeFlavor=mono` in some of the console text when building the test, what am I missing to run it as mono? (I'm assuming its not mono if i'm still seeing the output above'
```
Building managed test group 1: "/Users/mdhwang/runtime_droid/dotnet.sh" msbuild /Users/mdhwang/runtime_droid/src/tests/build.proj /t:Build "/p:TargetArchitecture=x64" "/p:Configuration=Debug" "/p:LibrariesConfiguration=Debug" "/p:TargetOS=Android" "/p:RuntimeOS=" "/p:RuntimeFlavor=mono" "/p:RuntimeVariant=" "/p:CLRTestBuildAllTargets=" "/p:__TestGroupToBuild=1" "/p:__SkipRestorePackages=1" /nodeReuse:false /maxcpucount /bl:/Users/mdhwang/runtime_droid/artifacts//log/Debug/InnerManagedTestBuild.1.binlog
  Microsoft (R) Build Engine version 17.0.0+c9eb9dd64 for .NET
  Copyright (C) Microsoft Corporation. All rights reserved.

  /Users/mdhwang/runtime_droid/.dotnet/sdk/6.0.100/MSBuild.dll -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/Users/mdhwang/runtime_droid/.dotnet/sdk/6.0.100/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/Users/mdhwang/runtime_droid/.dotnet/sdk/6.0.100/dotnet.dll -maxcpucount /maxcpucount -verbosity:m /bl:/Users/mdhwang/runtime_droid/artifacts//log/Debug/InnerManagedTestBuild.1.binlog /nodeReuse:false /p:TargetArchitecture=x64 /p:Configuration=Debug /p:LibrariesConfiguration=Debug /p:TargetOS=Android /p:RuntimeOS= /p:RuntimeFlavor=mono /p:RuntimeVariant= /p:CLRTestBuildAllTargets= /p:__TestGroupToBuild=1 /p:__SkipRestorePackages=1 /t:Build /Users/mdhwang/runtime_droid/src/tests/build.proj
  Building managed test group 2: "/Users/mdhwang/runtime_droid/dotnet.sh" msbuild /Users/mdhwang/runtime_droid/src/tests/build.proj /t:Build "/p:TargetArchitecture=x64" "/p:Configuration=Debug" "/p:LibrariesConfiguration=Debug" "/p:TargetOS=Android" "/p:RuntimeOS=" "/p:RuntimeFlavor=mono" "/p:RuntimeVariant=" "/p:CLRTestBuildAllTargets=" "/p:__TestGroupToBuild=2" "/p:__SkipRestorePackages=1" /nodeReuse:false /maxcpucount /bl:/Users/mdhwang/runtime_droid/artifacts//log/Debug/InnerManagedTestBuild.2.binlog
  Microsoft (R) Build Engine version 17.0.0+c9eb9dd64 for .NET
  Copyright (C) Microsoft Corporation. All rights reserved.

  /Users/mdhwang/runtime_droid/.dotnet/sdk/6.0.100/MSBuild.dll -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/Users/mdhwang/runtime_droid/.dotnet/sdk/6.0.100/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/Users/mdhwang/runtime_droid/.dotnet/sdk/6.0.100/dotnet.dll -maxcpucount /maxcpucount -verbosity:m /bl:/Users/mdhwang/runtime_droid/artifacts//log/Debug/InnerManagedTestBuild.2.binlog /nodeReuse:false /p:TargetArchitecture=x64 /p:Configuration=Debug /p:LibrariesConfiguration=Debug /p:TargetOS=Android /p:RuntimeOS= /p:RuntimeFlavor=mono /p:RuntimeVariant= /p:CLRTestBuildAllTargets= /p:__TestGroupToBuild=2 /p:__SkipRestorePackages=1 /t:Build /Users/mdhwang/runtime_droid/src/tests/build.proj
  Building managed test group 3: "/Users/mdhwang/runtime_droid/dotnet.sh" msbuild /Users/mdhwang/runtime_droid/src/tests/build.proj /t:Build "/p:TargetArchitecture=x64" "/p:Configuration=Debug" "/p:LibrariesConfiguration=Debug" "/p:TargetOS=Android" "/p:RuntimeOS=" "/p:RuntimeFlavor=mono" "/p:RuntimeVariant=" "/p:CLRTestBuildAllTargets=" "/p:__TestGroupToBuild=3" "/p:__SkipRestorePackages=1" /nodeReuse:false /maxcpucount /bl:/Users/mdhwang/runtime_droid/artifacts//log/Debug/InnerManagedTestBuild.3.binlog
  Microsoft (R) Build Engine version 17.0.0+c9eb9dd64 for .NET
  Copyright (C) Microsoft Corporation. All rights reserved.

  /Users/mdhwang/runtime_droid/.dotnet/sdk/6.0.100/MSBuild.dll -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,/Users/mdhwang/runtime_droid/.dotnet/sdk/6.0.100/dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,/Users/mdhwang/runtime_droid/.dotnet/sdk/6.0.100/dotnet.dll -maxcpucount /maxcpucount -verbosity:m /bl:/Users/mdhwang/runtime_droid/artifacts//log/Debug/InnerManagedTestBuild.3.binlog /nodeReuse:false /p:TargetArchitecture=x64 /p:Configuration=Debug /p:LibrariesConfiguration=Debug /p:TargetOS=Android /p:RuntimeOS= /p:RuntimeFlavor=mono /p:RuntimeVariant= /p:CLRTestBuildAllTargets= /p:__TestGroupToBuild=3 /p:__SkipRestorePackages=1 /t:Build /Users/mdhwang/runtime_droid/src/tests/build.proj
  ```

mdh1418 replied:
I'll remove all the associated `JitOptimizationSensitive` exclusions and check CI results again.

> Having that said it might be equal important to be less dependent on all these configs and since this is not an issue on MonoVM (not running tiered JIT), we could just make sure we run the tests independently, would it make sense to look at a more generic property, like runtime flavor when setting this property, that way we don't need to update these with a lot of different conditions when we start to run these tests on all Mono runtime platforms supporting them?

Ah, so instead conditioning `JitOptimizationSensitive` based on `RuntimeFlavor` in each csproj? 

> we could just make sure we run the tests independently

What would this look like?

mdh1418 replied:
> I'll remove all the associated `JitOptimizationSensitive` exclusions and check CI results again.

Lane `Build Android x64 Release AllSubsets_Mono_RuntimeTests` results https://helix.dot.net/api/jobs/50218d6c-c030-4a62-947e-7cfb705d399b/workitems/PayloadGroup0?api-version=2019-06-17, I see `SKIPPING EXECUTION BECAUSE COMPlus_TieredCompilation has not been disabled and this test is marked JitOptimizationSensitive
Test Harness Exitcode is : 0`

lateralusX replied:
> Ah, so instead conditioning `JitOptimizationSensitive` based on `RuntimeFlavor` in each csproj?

Yes, I believe not condition them to only Android but to MonoVM would make sense, alternative is to add the condition directly into the target that emits the script:

https://github.com/dotnet/runtime/blob/87777b624b0b8ea295b0fe9cd8505ce262ed75c0/src/tests/Common/CLRTest.Execute.Batch.targets#L96

https://github.com/dotnet/runtime/blob/87777b624b0b8ea295b0fe9cd8505ce262ed75c0/src/tests/Common/CLRTest.Execute.Bash.targets#L99

So adding something like `Condition="'$(RuntimeFlavor)' != 'mono'"` would make sure we run these tests on MonoVM independent on what CI configuration we use and what env variables that ends up setting.

lateralusX commented:
@mdh1418 How did you verify that the enabled tests now runs on Android on CI? Are we running the tests on all Android platforms currently tested on CI?

I believe this looks overall good, great to get the test running on mobile platforms.

mdh1418 commented:
> @mdh1418 How did you verify that the enabled tests now runs on Android on CI? Are we running the tests on all Android platforms currently tested on CI?
> 
> I believe this looks overall good, great to get the test running on mobile platforms.

I looked through the Android runtime tests lane on this PR's CI checks `runtime-extra-platforms` and in `PayloadGroup0`, I was able to find all of the tests in https://helix.dot.net/api/jobs/ed8baac3-e67c-47aa-95d8-0376b284f7a3/workitems/PayloadGroup0?api-version=2019-06-17 and looked through the adb log for a successful output of 100.
```
eventlistener
eventcounter
pollingcounter
regression-46938
regression-25709
incrementingpollingcounter
runtimecounters
incrementingeventcounter
gh53564
eventactivityidcontrol
```

However, I only checked that one lane so far `Build Android x64 Release AllSubsets_Mono_RuntimeTests`

mdh1418 committed:
Revert "[tests] Enable non-eventpipe Android tracing tests to emulate desktop mono"

This reverts commit 3b8ef2356a739ee66962bb65f81db9d8ae64f3f3.

mdh1418 commented:
/azp run runtime-extra-platforms

mdh1418 committed:
[tests] Skip JitOptimizationSensitive script on mono

mdh1418 committed:
Update Android GetStandardTransport TcpClient initialization

mdh1418 committed:
[tests] Skip diagnostics tests on browser and Android

mdh1418 commented:
/azp run runtime-extra-platforms

lateralusX commented:
@mdh1418 Can you verify that the last config in this PR now runs the test on Android? If so I believe this is good to go!

mdh1418 commented:
@lateralusX Just checked the lane `Build Android x64 Release AllSubsets_Mono_RuntimeTests` again, all the respective tests have a successful exit code of 100.
